core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala

```diff
@@ -122,6 +123,24 @@ private[ui] class StagePage(parent: StagesTab) extends WebUIPage("stage") {
       val externalAccumulables = allAccumulables.values.filter { acc => !acc.internal }
       val hasAccumulators = externalAccumulables.size > 0
 
+      val tasksLocality = new mutable.HashMap[String, Long]
+      if (stageData.taskData.size > 0) {
+        for ((key, value) <- stageData.taskData) {
+          val locality = value.taskInfo.taskLocality.toString;
+          if (tasksLocality.contains(locality)) {
+            val count: Long = tasksLocality.get(locality).get + 1;
+            tasksLocality.put(locality, count);
+          } else {
+            tasksLocality.put(locality, 1);
+          }
+        }
+      }
+      var localityLevels: String = "";
```

Comment:
How about re-writing this as: (all of the concat statements are tough to read)

tasksLocality.map { _ match {
  case (localityLevel, count) =>
    s"$localityLevel ($count)"
}}.mkString("; ")